warehouse/http.go

```diff
@@ -0,0 +1,415 @@
+		Handler:           bugsnag.Handler(srvMux),
+		ReadHeaderTimeout: a.config.readerHeaderTimeout,
+	}
+	return kithttputil.ListenAndServe(ctx, srv)
+}
+
+func (a *Api) addWarehouseEndpoints(ctx context.Context, mux *chi.Mux) {
+	a.logger.Infof("waiting for BackendConfig before starting on %d", a.config.webPort)
+
+	a.bcConfig.WaitForConfig(ctx)
+
+	// process the staging file
+	mux.Handle("/v1/process", (&api.WarehouseAPI{
+		Logger:      a.logger,
+		Stats:       a.statsFactory,
+		Repo:        a.stagingRepo,
+		Multitenant: a.tenantManager,
+	}).Handler())
+
+	// triggers upload only when there are pending events and triggerUpload is sent for a sourceId
+	mux.Post("/v1/warehouse/pending-events", a.pendingEventsHandler)
+
+	// triggers uploads for a source
+	mux.Post("/v1/warehouse/trigger-upload", a.triggerUploadHandler)
+
+	// Warehouse Async Job end-points
+	mux.Post("/v1/warehouse/jobs", a.asyncManager.InsertJobHandler)       // FIXME: add degraded mode
+	mux.Get("/v1/warehouse/jobs/status", a.asyncManager.StatusJobHandler) // FIXME: add degraded mode
+
+	// TODO: Remove this endpoint once sources change is released
+	// fetch schema info
+	mux.Get("/v1/warehouse/fetch-tables", a.fetchTablesHandler)
+	mux.Get("/internal/v1/warehouse/fetch-tables", a.fetchTablesHandler)
+}
+
+func (a *Api) healthHandler(w http.ResponseWriter, r *http.Request) {
+	var dbService, pgNotifierService string
+
+	ctx, cancel := context.WithTimeout(r.Context(), a.config.healthTimeout)
+	defer cancel()
+
+	if a.config.runningMode != DegradedMode {
+		if !checkHealth(ctx, a.notifier.GetDBHandle()) {
+			http.Error(w, "Cannot connect to pgNotifierService", http.StatusInternalServerError)
+			return
+		}
+		pgNotifierService = "UP"
+	}
+
+	if isMaster(a.mode) {
+		if !checkHealth(ctx, a.db.DB) {
+			http.Error(w, "Cannot connect to dbService", http.StatusInternalServerError)
+			return
+		}
+		dbService = "UP"
+	}
+
+	healthVal := fmt.Sprintf(`
+{
+	"server": "UP",
+	"db": %q,
+	"pgNotifier": %q,
+	"acceptingEvents": "TRUE",
+	"warehouseMode": %q
+}
+	`,
+		dbService,
+		pgNotifierService,
+		strings.ToUpper(a.mode),
+	)
+
+	_, _ = w.Write([]byte(healthVal))
+}
+
+func checkHealth(ctx context.Context, db *sql.DB) bool {
+	if db == nil {
+		return false
+	}
+
+	healthCheckMsg := "Rudder Warehouse DB Health Check"
+	msg := ""
+
+	err := db.QueryRowContext(ctx, `SELECT '`+healthCheckMsg+`'::text as message;`).Scan(&msg)
+	if err != nil {
+		return false
+	}
+
+	return healthCheckMsg == msg
+}
+
+// pendingEventsHandler check whether there are any pending staging files or uploads for the given source id
+func (a *Api) pendingEventsHandler(w http.ResponseWriter, r *http.Request) {
+	a.logger.LogRequest(r)
+
+	defer func() { _ = r.Body.Close() }()
+
+	var pendingEventsReq pendingEventsRequest
+
+	if err := json.NewDecoder(r.Body).Decode(&pendingEventsReq); err != nil {
+		a.logger.Errorf("invalid JSON body for pending events: %v", err)
```

Comment:
[minor] structured logging and warning should be more suitable
```suggestion
		a.logger.Warnw("invalid JSON body for pending events", log.Error, err)
```